### DISCUSSION TITLE
Can't upload a ROMS file with FieldSet

### DISCUSSION THREAD

**Discussion by micgros** [Parcels-code/Parcels#1492] [Q&A] [answered]
Dear Community,
I have a Parcels simulation that so far I ran with a FOCI output. Unfortunately, FOCI's resolution is not high enough for our purposes, so I got a ROMS output with a better resolution. Since ROMS has a C grid, I tried to run it with `from_c_grid_dataset`, but it doesn't work. I also followed every possible thread and tutorial, tried every possible variable combination,  but I couldn't get to the right syntax to read this file. I hope someone can help with uploading it to my script. Here is the list of variables from the file:

```
Dimensions:     (s_rho: 1, eta_rho: 272, xi_rho: 456, eta_u: 272, xi_u: 455,
                 eta_v: 271, xi_v: 456, ocean_time: 124)
Coordinates:
    lat_rho     (eta_rho, xi_rho) float64 ...
    lat_u       (eta_u, xi_u) float64 ...
    lat_v       (eta_v, xi_v) float64 ...
    lon_rho     (eta_rho, xi_rho) float64 ...
    lon_u       (eta_u, xi_u) float64 ...
    lon_v       (eta_v, xi_v) float64 ...
    ocean_time  (ocean_time) datetime64[ns] 2008-03-01T03:00:00 ... 2008-03-3...
    s_rho       (s_rho) float64 -0.01667
Dimensions without coordinates: eta_rho, xi_rho, eta_u, xi_u, eta_v, xi_v
Data variables:
    Cs_r        (s_rho) float64 ...
    h           (eta_rho, xi_rho) float64 ...
    hc          float64 ...
    u           (ocean_time, s_rho, eta_u, xi_u) float32 ...
    v           (ocean_time, s_rho, eta_v, xi_v) float32 ...
    zeta        (ocean_time, eta_rho, xi_rho) float32 ...
Attributes: (12/38)
    file:              /nesi/nobackup/niwa00020/collinsc/work/hikurangi/sim24...
    format:            netCDF-3 64bit offset file
    Conventions:       CF-1.4, SGRID-0.3
    type:              ROMS/TOMS nonlinear model averages file
    title:             ROMS - NZ Shelf
    var_info:          varinfo.dat

```

**Comment by erikvansebille** [comment]
Dear @micgros, thanks for asking this question. We don't use ROMS output in my team here so Im not sure I can help. But I can imagine that it would be easier to advise you if you also provide the code you now use to create a FieldSet, and explain _what_ goed wrong. Why does your code not work?

**Reply by micgros** [reply]
Dear @erikvansebille, thank you for your reply. 

As I wrote, I tried A LOT of different code versions to read the file. 

Here are some:

```
# Set up the FieldSet
filenames = {
    'U': 'uvSurf_2008_03.nc',
    'V': 'uvSurf_2008_03.nc'
}
variables = {'U': 'u', 'V': 'v'}
dimensions = {
    'U': {'lon': 'lon_rho', 'lat': 'lat_rho', 'depth': 'depth_rho', 'time': 'ocean_time'},
    'V': {'lon': 'lon_rho', 'lat': 'lat_rho', 'depth': 'depth_rho', 'time': 'ocean_time'}
}

fieldset = FieldSet.from_c_grid_dataset(filenames, variables, dimensions, mesh='flat')
```

I get this error:
ValueError: all the input array dimensions except for the concatenation axis must match exactly, but along dimension 3, the array at index 0 has size 455 and the array at index 1 has size 456

If I change the coordinates to: `'lon': 'lon_u', 'lat': 'lat_u'`.... I get this error:
ValueError: On a C-grid, the dimensions of velocities should be the corners (f-points) of the cells, so the same for U and V. See also ../examples/documentation_indexing.ipynb

```
# Load the dataset
ds = xr.open_dataset('uvSurf_2008_03.nc')

# Pad the U field (add an extra column)
u_padded = xr.concat([ds['u'], xr.zeros_like(ds['u'].isel(xi_u=-1))], dim='xi_u')
u_padded.to_netcdf('u_padded.nc')

# Pad the V field (add an extra row)
v_padded = xr.concat([ds['v'], xr.zeros_like(ds['v'].isel(eta_v=-1))], dim='eta_v')
v_padded.to_netcdf('v_padded.nc')

# Now, create the FieldSet with the padded data
from parcels import FieldSet

filenames = {
    'U': 'u_padded.nc',
    'V': 'v_padded.nc'
}
variables = {'U': 'u_padded', 'V': 'v_padded'}
dimensions = {
    'U': {'lon': 'lon_rho', 'lat': 'lat_rho', 'time': 'ocean_time'},
    'V': {'lon': 'lon_rho', 'lat': 'lat_rho', 'time': 'ocean_time'}
}

fieldset = FieldSet.from_c_grid_dataset(filenames, variables, dimensions, mesh='flat')

```
I get this error:
KeyError: 'lon_rho'

I also suspect that the file might has a conflict with dimensions. I already contacted the person who gave me this file, but meanwhile I try also here :) Many thanks!

**Reply by erikvansebille** [reply]
I don't have access to your fields so can't test, but you could try

```python
variables = {'U': 'u', 'V': 'v'}
dimensions = {
    'U': {'lon': 'xi_rho', 'lat': 'eta_rho', 'depth': 'depth_rho', 'time': 'ocean_time'},
    'V': {'lon': 'xi_rho', 'lat': 'eta_rho', 'depth': 'depth_rho', 'time': 'ocean_time'}
}

fieldset = FieldSet.from_c_grid_dataset(filenames, variables, dimensions)
```

This way, you use the grid point information on the zeta-points. This is similar to the f-points in the NEMO-grid (I think...)

**Reply by micgros** [reply]
Hi Erik,

Thanks! It still doesn't work, I get the following error again:

ValueError: all the input array dimensions except for the concatenation axis must match exactly, but along dimension 3, the array at index 0 has size 455 and the array at index 1 has size 456

I think that this file might has a dimension conflict. I already contacted the guy who gave me this file and he will try to help. I'll update here on the thread, but if someone has another idea, that will be awesome!

**Reply by erikvansebille** [reply]
Ah, perhaps then add the following?

```python
indices = {'lon': range(455), 'lat': range(271)}
fieldset = FieldSet.from_c_grid_dataset(filenames, variables, dimensions, indices=indices)
```

**Reply by micgros** [reply]
Hi Erik, many thanks! It worked!! However, we are now getting many different errors (not from FieldSet).
For example, when we run the script as you suggested (only without `'depth': 'depth_rho'` as the file is 2D (we have a variable name `zeta` that I suspect that has something to do with depth)), we get this error:

 ```
File "/Users/michal/anaconda3/envs/parcels2/lib/python3.12/site-packages/parcels/kernel.py", line 585, in execute
    raise FieldOutOfBoundError(p.lon, p.lat, p.depth)
parcels.tools.statuscodes.FieldOutOfBoundError: Field sampled out-of-bound, at (172.9972981266724, -45.99935258501933, 0.0)
```
However, when I add to the dimensions: `'depth': 'zeta'` or `'zeta': 'zeta'`, we get:
```
 File "/Users/michal/anaconda3/envs/parcels2/lib/python3.12/site-packages/parcels/field.py", line 1512, in __init__
    assert self._check_grid_dimensions(U.grid, V.grid), (
AssertionError: Dimensions of U and V are not the same.
```

**Reply by erikvansebille** [reply]
If you want 2D simulations (so at the surface), you should not include `depth` in the `dimensions` dictionary at all. So
```python
dimensions = {
    'U': {'lon': 'xi_rho', 'lat': 'eta_rho', 'time': 'ocean_time'},
    'V': {'lon': 'xi_rho', 'lat': 'eta_rho', 'time': 'ocean_time'}
}
```
Parcels will then assume that there is no depth dimensions, and particles will move horizontally using the first depth level

**Reply by JamiePringle** [reply]
Just to jump in as a long time ROMS user. Zeta is NOT the depth, it is the free surface height, which can be negative!  Also, note that ROMS grids, as you would expect for a naive C-grid implementation, do not have the same dimensions for any of the variables U, V, or Rho/T. Sketch out a c-grid and you can see why. It is actually quite odd that Nemo has the same size for all of its grids. A good picture to stare at can be found at 
https://www.myroms.org/wiki/File:Whole_grid.png . Note how the U grid is larger in the x direction, V in y, and Rho/T in neither.

**Reply by micgros** [reply]
Hi @JamiePringle, thank you for all the insights on ROMS. First of all, thank you for pointing out on zeta. I'm not using it as I'm running a 2D simulation. This one is relevant also to @erikvansebille, the only reason I tried to add zeta is because without any depth dimension input argument, as Erik pointed out to me on his last comment: `dimensions = {'lon': 'xi_rho', 'lat': 'eta_rho', 'time': 'ocean_time'}`, I get an error:
```
File "/Users/michal/anaconda3/envs/parcels2/lib/python3.12/site-packages/parcels/kernel.py", line 585, in execute
   raise FieldOutOfBoundError(p.lon, p.lat, p.depth)
parcels.tools.statuscodes.FieldOutOfBoundError: Field sampled out-of-bound, at (172.9972981266724, -45.99935258501933, 0.0)
```
The error includes depth (FieldOutOfBoundError(p.lon, p.lat, **p.depth**) and out-of-bound at (x, y, **0.0**). This is why I checked what happens if I add a depth dimension argument (there is no depth dimension in the file I'm using, only that zeta that has meter units and irrelevant as you said). 

And you are right, the dimensions of U are: `455 272 124` while V's are:`456 271 124`. But then I'm not so sure whether @erikvansebille's way to read this file, using the `indices=indices` (see above), is correct? Any suggestion?

**Comment by JamiePringle** [comment]
Alas, no, I have not tried to track particles in ROMS fields. I know Rob
Hetland and others have tried.

**Reply by micgros** [reply]
@JamiePringle Thanks anyway :)

**Comment by nitsanavni** [comment]
Thanks @JamiePringle !

I'm working with @micgros on loading a ROMS file and using it with `parcels`.

@hetland if we could get your help on this, we would sure appreciate it!

**Reply by hetland** [reply, marked as answer]
The most recent think I've done with Parcels is in [this script](https://github.com/hetland/sunrise/blob/master/Scripts/Google_earth_sss_drifters.py)

It's not new, and only 2D surface flow, so you may have to make some changes to make it work the way you want. Hopefully not too many changes are required.

**Reply by nitsanavni** [reply]
Thanks so much! We'll give it a go.

**Reply by micgros** [reply]
@hetland It worked!! Thank you very much for sharing the script! Using `xarray` worked!
